Migration tool

karottenreibe opened the pull request.

nihathrael committed:
Add Pylint report format

nihathrael committed:
Update to 3.8

nihathrael committed:
Attempt to fix travis-sphinx build

karottenreibe committed:
enable new movetolastcommit parameter for report uploads

nihathrael committed:
Merge pull request #25 from cqse/coverage_upload_parameters

enable new movetolastcommit parameter for report uploads

alexrhein committed:
added path suffix to connector config

nihathrael committed:
Fix whitespaces

nihathrael committed:
Merge pull request #26 from alexrhein/path_suffix_in_connector_config

added path suffix to connector config

nihathrael committed:
Prepare for 3.9.0

Macariel committed:
Added a migration tool (works for tasks and blacklist)

Macariel committed:
Fixed some typos

karottenreibe committed:
fix typo

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+from __future__ import absolute_import
```
please rename the file to something more specific, e.g. blacklist_and_finding_migration etc.

karottenreibe replied:
General remark: this script will stop working as soon as we have the branch-specific blacklists, right? please discuss with Martin if this is intended or if we need to support them

karottenreibe replied:
please add a shebang to the file and make it executable so people can call it like ./migration.py directly
please add a comment at the beginning of this file, explaining what it does and how to use it

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+}
+
+
+def main():
```
read arguments from the command line or a config file? you can use one of the many python command line parsers if you like or just use positional arguments. I just think that hardcoding the values is not the best solution since I can't chekcout and use the program, I have to modify it every time

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+    """ Base class for migrating data from one instance to another via REST calls. """
+
+    def __init__(self, old, new):
+        self.old = old
```
please use more specific variable names. it's unclear what old and new refers to

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+    def get(client, service, path_param="", parameters=None):
+        """ Performs a GET call from the client to the service with the given parameters
+        and returns the response as a JSON Object.
+        The path parameter are additions to the path, e.g /service/id.
```
I don't understand this part of the comment. you are passing the service name into this method separately already, so why would you ever want to add it again as a suffix? sohuldn't the example be just /id or something like that?

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+        The path parameter are additions to the path, e.g /service/id.
+        """
+        response = client.get(client.get_project_service_url(service) + path_param, parameters)
+        return json.loads(response.text)
```
please add code to handle error responses (e.g. 404, 500) from the server. not sure if that should abort the migration or not. please discuss with martin

Macariel replied:
It can be done. At the moment though, I abort everytime a ServiceError is raised. I think we need to check at the MR what problems might arise and if there is actually a use case in which resuming the migration with a 404 or 500 make sense. It would add a lot of None checks throughout the migrators and I rather not do this preemptively, as it makes the code uglier.

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+        """
+        return self.get(self.new, service, path_param, parameters)
+
+    def put(self, service, data, path_param="", parameters=None):
```
please rename to put_to_new to be consistent with the other methods above

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+    @staticmethod
+    def match_finding(finding1, finding2):
+        """ Checks if the given two findings are the same. """
+        if finding1["message"] != finding2["message"]:
```
this might break if we migrate between different versions of Teamscale (as messages may change in between). Is that a supported feature? if not, please indicate so somewhere, maybe even check the TS versions before starting the migration?

karottenreibe committed:
fix typo

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+
+    def migrate(self):
+        """ Migrates the blacklists. """
+        migrate_blacklist = self.get_blacklist_infos()
```
rename to blacklist_to_migrate to be clearer?

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+    def migrate(self):
+        """ Migrates the blacklists. """
+        migrate_blacklist = self.get_blacklist_infos()
+        for blacklisted in migrate_blacklist:
```
rename to blacklisted_finding to be clearer?

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+        migrate_blacklist = self.get_blacklist_infos()
+        for blacklisted in migrate_blacklist:
+            new_id = self.get_matching_finding_id(blacklisted["findingId"])
+            if not new_id:
```
are you sure that if not X will only match the cases we want? how about if X is not None instead. AFAICS None is the only value we really want to check here, right?

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+        for blacklisted in migrate_blacklist:
+            new_id = self.get_matching_finding_id(blacklisted["findingId"])
+            if not new_id:
+                print("could not find finding %s in new Teamscale" % blacklisted["findingId"])
```
if possible, please print not just finding IDs but links to the findings in error messages. that makes it easier for the user to look into issues
please mark error messages so they are easily identifed among the rest of the messages. E.g. use some logging framework

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+    """
+
+    def migrate(self):
+        """ Migrates the blacklists. """
```
why blacklists (plural)?
also, please document that this will skip already migrated entries

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+        infos = []
+        for finding_id in blacklisted_ids:
+            info = self.get_from_old("finding-blacklist", path_param=finding_id)
+            if not info:
```
are you sure this actually works? I think the server will send a 404 response (which is not valid json) which you don't handle specifically above in #get() so I think this might just throw some exception if the blacklist entry can't be found

Macariel replied:
Well, there no reason why this shouldn't find anything tbh.
The ids I am getting the info for where retrieved with the same service call.
Calling "finding-blacklist" without arguments gives you all ids. And then I am calling the same service with the ids.
If this fails, there is something really wrong and it actually should fail

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+    def migrate(self):
+        """ Migrates the tasks. """
+        parameters = {"with-count": True, "max": 0}
+        new_tasks_count = self.get_from_new("tasks", parameters=parameters)["totalTaskCount"] + 1
```
this will fail if the service returns e.g. 500 or 404

karottenreibe replied:
why +1? please add an inline comment

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+
+        old_tasks = self.get_from_old("tasks", parameters={"details": True})
+        total = len(old_tasks)
+        for old_task in old_tasks:
```
this loop is a bit convoluted. I'd prefer it if you separated it into two parts: first is filtering the old tasks based on whether they already exist, second is migrating. this way, you can also calculate the total with a simple len() call after filtering, which is much easier to understand

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+        If any finding cannot be matched on the new instance `False` will be returned, `True` otherwise."""
+        for finding in task["findings"]:
+            matching_finding_id = self.get_matching_finding_id(finding["findingId"])
+            if not matching_finding_id:
```
same as above: I'd check for X is not None

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+
+    def task_exists(self, old_task):
+        """ Checks if the given tasks already exists on the new instance. """
+        new_tasks = self.get_from_new("tasks", parameters={
```
it seems expensive to retreive all tasks over and over again when checking if old task exists. Can we cache this, e.g. by moving it out of this method?

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+        return self.str_task(task1) == self.str_task(task2) and len(task1["findings"]) == len(task2["findings"])
+
+    @staticmethod
+    def str_task(task):
```
why convert the array to a string? you can compare arrays directly

karottenreibe commented on tools/migration.py:
```diff
@@ -0,0 +1,248 @@
+
+        for old_finding in old_findings:
+            match_found = False
+            for new_finding in new_findings:
```
extract a method for this loop? that way, you can get rid of the match_found variable

Macariel committed:
Reworked the migrating tools

Macariel committed:
Added warning and mitigations for migrating between different TS-Versions

Macariel committed:
Added an exemplary test

Macariel committed:
Fixed some findings

Macariel committed:
Changed module name

Macariel committed:
Added fallback for missing service call

Macariel committed:
Disabled api check. Not working for 3.6 even though 3.2 should be fine

Macariel committed:
Revert Change

Macariel committed:
Changed module import

Macariel committed:
Revert "Changed module import"

This reverts commit 4b0df559179178ef75f857dbb1983d675cbde44f.

Macariel committed:
Revert "Revert Change"

This reverts commit eee05e6be74b585990f4d4fc339831a0c07b8aec.

Macariel committed:
Revert "Disabled api check. Not working for 3.6 even though 3.2 should be fine"

This reverts commit 99f40667210f47a4cf500a4b9da8ba3bef667d89.

Macariel committed:
Revert "Added fallback for missing service call"

This reverts commit 55ffb8154fd3032ceb007a4f87340ee204be8d8a.

Macariel committed:
Changed import and added pytest cache to gitignore

Macariel committed:
Added handling of missing findings

Macariel committed:
Fixed bug

Macariel committed:
Added a dry-run option

Macariel committed:
Added more logs

Macariel committed:
Added logging level print

Macariel committed:
Added more logging

Macariel committed:
Fixed id

Macariel committed:
Logging

Macariel committed:
Modified logging

Macariel committed:
Fixed wrong method call

Macariel committed:
Changed the task migration

Macariel committed:
Stringified path suffix

Macariel committed:
Changed the task migration

Macariel committed:
Changed comparison of tasks

Macariel committed:
Removed task pre-filtering

Macariel committed:
Added the step by step option

Macariel committed:
Removing print statement

Macariel committed:
Removed dead code and added a few comments

Macariel committed:
Added a check if a project exists

Macariel committed:
Added a scripts which allows the batch migration of blacklists and tasks

Commit:
Changed output and made it clearer

Macariel committed:
Merge branch 'migration-tool' of https://github.com/cqse/teamscale-client-python into migration-tool

Macariel committed:
Added logging%

Macariel committed:
Merge branch 'migration-tool' of https://github.com/cqse/teamscale-client-python into migration-tool

Macariel committed:
Fixed the double logging

Macariel committed:
Merge branch 'migration-tool' of https://github.com/cqse/teamscale-client-python into migration-tool

mpdeimos committed:
Merge branch 'master' into migration-tool

mpdeimos commented on tools/migration/blacklist_migrator.py:
```diff
@@ -0,0 +1,57 @@
+        blacklist_infos = self.get_blacklist_infos()
+        if len(blacklist_infos) == 0:
+            self.logger.info("No new blacklisted findings to migrate")
+            exit(1)
```
Do we have to exit hard here?

Macariel replied:
Fixed

mpdeimos commented on tools/migration/batch_config.template:
```diff
@@ -0,0 +1,16 @@
+        "token": "tokentoken"
+    },
+    "project_mappings": [
+        {"from": "a", "to": "b"},
```
use simple json hashmaps, e.g {"a": "b", "c": "d"}?

Macariel replied:
I actually did that on purpose to prevent any kind of confusion from where to where the migration is happening

mpdeimos replied:
makes sense

mpdeimos commented on tools/migration/migrator_base.py:
```diff
@@ -0,0 +1,242 @@
+            logger.exception("Config file '%s' is malformed" % path, exc_info=True)
+    else:
+        logger.exception("Config file '%s' does not exist" % path)
+    exit(1)
```
does this stop the migrator?

Macariel replied:
Yes

mpdeimos commented on tools/migration/migrator_base.py:
```diff
@@ -0,0 +1,242 @@
+    def check_projects(self):
+        """ Check if the two project actually do exist on the given servers. """
+        try:
+            check_url = "{0.url}/create-project/{0.project}"
```
I think the correct call for this is /projects/{id}, see https://git.cqse.eu/cqse/teamscale/blob/master/lib/Cqse.Teamscale.Client/src/Service/Implementation/ProjectService.cs

Macariel replied:
Fixed

mpdeimos commented on tools/migration/migrator_base.py:
```diff
@@ -0,0 +1,242 @@
+                response = client.get(url, parameters).json()
+            except ServiceError as e:
+                status_code = e.response.status_code
+                if status_code == 500:
```
Any reason why we handle 500 special here?

Macariel replied:
I am not sure anymore, actually. It probably was important at some point. I deleted a lot of code, so it was probably was connected with that. Will change it

mpdeimos commented on tools/migration/blacklist_migrator.py:
```diff
@@ -0,0 +1,57 @@
+                self.logger.warning("Could not match finding %s to new instance" % self.get_findings_url(old_id))
+            else:
+                self.logger.info("Migrating blacklisted finding %s" % self.get_findings_url(old_id))
+                self.blacklist_finding(blacklist_info, new_id)
```
check if already blacklisted?

Macariel replied:
I removed that explicitly.
Often I was encountering the following problem.
Looking at the UI, a project had, for example, 18 blacklisted findings. Upon migrating the blacklist the result might have said something like: 6/34 findings where successfully blacklisted.
Now I wasn't sure ift there was actually a problem or not. When I removed the filtering of already blacklisted findings it said: 18/34 findings where blacklisted. Upon closer inspection I found out, that the other 16 findings did no longer exist on any project there.
So I kept it, because if the numbers where actually similar it is easier to determine if the migration was successful or not. We are not really running into any performance bottleneck here, so in the end it doesn't really matter.

mpdeimos replied:
makes sense

mpdeimos reviewed (changes requested):
Please see comments for details. Any chance to get the build passing?

Macariel committed:
Changed according to review

Macariel committed:
Added path prefix transformation

Macariel committed:
Deleted the wrong config file

Macariel committed:
Changed findings comparison

Macariel committed:
Path transformation fix

Macariel committed:
Removed version check

Macariel committed:
Made findings comparision more robust

Macariel committed:
Added a description on how to use the migration tools

Macariel committed:
tmp

Macariel committed:
Changed the findings comparison

Macariel committed:
Reset client to master

Macariel committed:
Small fixes

Macariel committed:
Added upload of finding-groups and -descriptions

r2h2 committed:
Path prefix transformation is defined in the 'new_instance'.
Additionally, fixed some linting errors (which resulted in unhandled
exceptions during task migration)

Macariel committed:
Merge pull request #37 from cqse/fix_path_prefix_transformation

Path prefix transformation is defined in the 'new_instance'.

Macariel committed:
Adjust README

Macariel committed:
Merge branch 'master' into migration-tool

Macariel committed:
Simplified exception logging

Commit:
Merge branch 'master' into migration-tool

Commit:
Fixes

Commit:
Add socks proxy support

Commit:
Improvements

ke-kx committed:
manually generate the correct service url when fetching tasks

ke-kx committed:
Merge branch 'migration-tool' into get_tasks_dirty_fix

ke-kx committed:
workaround for posting task comments

Commit:
add blacklisted parameter to get_findings to enable fetching false positives

albertsteckermeier committed:
Merge branch 'master' into migration-tool

Commit:
add functionality for querying findings counts and descriptions

Commit:
remove unnecessary proxy parameter

Commit:
Merge branch 'master' into a_team_branch

Commit:
Merge branch 'a_team_branch' into migration-tool

Commit:
Support code for ongoing work on temp findings script

Commit:
Added new supporting methods to get finding churn by commit timestamp

Commit:
Support code for ongoing work on temp findings script

Commit:
Refactoring the solution and changing output

Commit:
Fixed a bug

baralCqse committed:
test gap analysis work

Commit:
Added helper method in merge_request data class
